File: java/src/main/java/io/github/skunkwerkx/hyperuuid/RfcBytes.java
package io.github.skunkwerkx.hyperuuid;

import java.util.UUID;

/**
 * Converts between {@link UUID} and its 16 RFC 9562 network-byte-order bytes.
 * {@code UUID}'s {@code getMostSignificantBits}/{@code getLeastSignificantBits} already
 * decompose into this exact byte order, so this is just two shift loops — no swapping,
 * unlike .NET's {@code Guid}.
 */
final class RfcBytes {
    private RfcBytes() {}

    static UUID fromRfcBytes(byte[] bytes) {
        return fromRfcBytes(bytes, 0);
    }

    /** The UUID whose 16 RFC 9562 bytes start at {@code offset} in {@code bytes}. */
    static UUID fromRfcBytes(byte[] bytes, int offset) {
        long msb = 0;
        for (int i = offset; i < offset + 8; i++) {
            msb = (msb << 8) | (bytes[i] & 0xFF);
        }
        long lsb = 0;
        for (int i = offset + 8; i < offset + 16; i++) {
            lsb = (lsb << 8) | (bytes[i] & 0xFF);
        }
        return new UUID(msb, lsb);
    }

    static byte[] toRfcBytes(UUID uuid) {
        byte[] bytes = new byte[16];
        long msb = uuid.getMostSignificantBits();
        for (int i = 7; i >= 0; i--) {
            bytes[i] = (byte) (msb & 0xFF);
            msb >>>= 8;
        }
        long lsb = uuid.getLeastSignificantBits();
        for (int i = 15; i >= 8; i--) {
            bytes[i] = (byte) (lsb & 0xFF);
            lsb >>>= 8;
        }
        return bytes;
    }
}

File: kotlin/src/main/kotlin/io/github/buvinghausen/hyperuuid/RfcBytes.kt
package io.github.buvinghausen.hyperuuid

import java.util.UUID

/**
 * Converts 16 RFC 9562 network-byte-order bytes into a [UUID]. `java.util.UUID`'s
 * `mostSignificantBits`/`leastSignificantBits` already decompose into this exact byte order,
 * so this is just two shift loops — no swapping, unlike .NET's `Guid`.
 */
internal fun uuidFromRfcBytes(bytes: ByteArray): UUID {
    var msb = 0L
    for (i in 0 until 8) msb = (msb shl 8) or (bytes[i].toLong() and 0xFF)
    var lsb = 0L
    for (i in 8 until 16) lsb = (lsb shl 8) or (bytes[i].toLong() and 0xFF)
    return UUID(msb, lsb)
}

/** Converts a [UUID] into its 16 RFC 9562 network-byte-order bytes. */
internal fun UUID.toRfcBytes(): ByteArray {
    val bytes = ByteArray(16)
    var msb = mostSignificantBits
    for (i in 7 downTo 0) {
        bytes[i] = (msb and 0xFF).toByte()
        msb = msb ushr 8
    }
    var lsb = leastSignificantBits
    for (i in 15 downTo 8) {
        bytes[i] = (lsb and 0xFF).toByte()
        lsb = lsb ushr 8
    }
    return bytes
}
